### PR TITLE
[WIP] implementing transolve

### LINKED ISSUE #11307: solveset can't solve 2**x-4
Solveset is unable to solve 2**x-4
```
>>> solveset(2**x-4)
ConditionSet(x, Eq(2**x - 4, 0), Complexes((-oo, oo) x (-oo, oo), False))
```

### PR DESCRIPTION
**Pull request by Yathartha22**
In this PR I have tried to implement `transolve` to make it work for `logarithmic and exponential equations`.
Initial commit was made in [PR](https://github.com/sympy/sympy/pull/11540). I have modified and reimplemented a bit so that it works and also added few `test cases`.

- [x] logarithmic equations
- [x] exponential equations
- [ ] equations solvable by LambertW function
- [ ] implementing trigonometric equation solver
 
ping @asmeurer @smichr @jksuom  @aktech @Shekharrajak @kshitij10496 

fixes: #11307

### CODE REVIEW

**Comment by aktech** [comment]
Please cherry-pick the commit from #11540 to retain authorship, instead of a new commit.

**Comment by Yathartha22** [comment]
I haven't used cherry-pick before, could you please help me with this.
I followed this [answer](https://stackoverflow.com/a/1405189/6446166) of StackOverflow as a reference.
[This](https://github.com/Yathartha22/sympy/blob/test/sympy/solvers/solveset.py#L791) is what the changes look like. I am not sure about it.

**Comment by Yathartha22** [comment]
@aktech I have made the changes as you suggested. Please have a look.

File: sympy/solvers/tests/test_solveset.py [hunk 1]
```diff
@@ -551,19 +555,34 @@ def test_poly_gens():
     assert solveset_real(4**(2*(x**2) + 2*x) - 8, x) == \
         FiniteSet(-Rational(3, 2), S.Half)
 
+def test_expo():
+    x = Symbol('x')
+    assert solveset(4**(5-9*x) - 8**(2-x), x, S.Reals) == FiniteSet(Rational(4, 15))
```
Comment (Shekharrajak) [inline sympy/solvers/tests/test_solveset.py]: Can you write test cases for complex solutions as well ?

Comment (Yathartha22) [inline sympy/solvers/tests/test_solveset.py]: Well, as of now I haven't made it to work for complex domain. I will be making few changes in `_invert_complex` and then add the test cases.

**Comment by Yathartha22** [comment]
@Shekharrajak I have added few test cases for complex solutions as well. Please have a look.

**Comment by Yathartha22** [comment]
@smichr @asmeurer @jksuom

**Comment by Yathartha22** [comment]
#11096 should be taken care while implementing exponential solver.

**Comment by aktech** [comment]
@Yathartha22 Can you please fix conflicts?

File: sympy/solvers/solveset.py [hunk 2]
```diff
@@ -149,6 +149,32 @@ def _invert_real(f, g_ys, symbol):
         if g is not S.Zero:
             return _invert_real(h, imageset(Lambda(n, n - g), g_ys), symbol)
 
+        # if there are two terms with g_ys=0
+        # e.g. 3*f(x) + 2*g(x) -> f(x)/g(x) = -2/3,
+        # where f(x) and g(x) are power terms
+        if not g and not g_ys.args[0] and len(h.args) == 2 and \
```
Comment (aktech) [inline sympy/solvers/solveset.py]: In a system like solvers where almost everything is heuristics, these conditions don't makes sense easily and would slowly get out of hand after multiple iterations of code changes over time. This is one of the reasons why `solve` is hard to understand.

To handle this situation, I would suggest to first document the what, why and how each and every condition and its proof of correctness, that would help in further development and extension. You also need to write meaningful functions wherever there is an independent piece/logic of code, which may or may not change to incorporate changes/addition of more conditions or situations to handle more equations.

**Review by aktech — CHANGES_REQUESTED** [review]

**Comment by aktech** [comment]
Closing in favor of #14736

**Pull request closed without merging**